### ISSUE TITLE
[Question/Clarification] - Ignoring gems when using Zeitwerk

### ISSUE THREAD

**Issue by sanket0354** [fxn/zeitwerk#149]
This is more of a Clarification/Question on what works or what is the right way to make the below work.

We have the Following Situation:

Gem A (Sort of like a Rails Engine) --used by-->  Gem B (Sort of like a Rails Engine) --used by--> Rails Application

Gem A has `concerns` as a namespace. It also has some nested directory structure (Basically nested concerns structure in models):
```
---app
  --- models
     --- concerns
        ---file1.rb
     --- folder1
        ---concerns
           ---filex.rb
     --- folder2
        ---concerns
           ---filey.rb
```
My questions:
- Can we ignore gems completely from auto-loading The rails application keeps complaining about concerns in Gem A. Is that even possible to get our rails application to use zeitwerk and not worry about gem?

- In Case when the above is not possible, Will the refactoring of `concerns` be sufficient to get the application upgraded even with the nested structure as mentioned above?

**Comment by fxn** [comment]
If Gem A is not an engine (only sort of), who is configuring its directories as autoload paths?

**Comment by fxn** [comment]
Which constant does `app/models/concerns/folder1/concerns/filex.rb` define?

**Comment by sanket0354** [comment]
Hey, Apologies again for the slight indentation mistake while creating the directory structure. I changed it above in the description a bit.

So the constant that `app/models/concerns/folder1/concerns/filex.rb`  defines will be:
 ```
 module Folder1::Concerns
    module Filex extend ActiveSupport::Concern
 ```

**Comment by fxn** [comment]
This should work out of the box. Which error do you get?

**Comment by sanket0354** [comment]
Right, will it? So you mean without refactoring anything in regards to concerns and also using zeitwerk in the rails app as it's inside the `models` directory? 

And I can use the concerns like `include ::Folder1::Concerns::Filex` 

Or will it be more like? `include ::Folder1::Filex`

**Comment by fxn** [comment]
Right, should work, and `include ::Folder1::Concerns::Filex` as you have it is correct.

The subdirectory `app/model/concerns` is special because that one is configured to be a root directory by Rails. So, you cannot have `Concerns::Folder1`, it has to be `Folder1`. But the good news is that your engine does not have a leading `Concerns` namespace, so you are all set with respect to that.

Then, "concerns" is not special anywhere else, it can be used as a nested namespace.

**Comment by fxn** [comment]
To put it in perspective, the name of a root directory is irrelevant, it represents the root namespace `Object`.

So, you cannot have `Concerns::Folder` by the same reason you cannot have `Models::User`, or `Controllers::UsersController`. `app/models` is a root directory, and `app/models/concerns` is a root directory too, they represent `Object`.

In descendants of root directories (that are not root directories themselves), names are arbitrary. There, "concerns" is no different than "foo".

**Comment by sanket0354** [comment]
I see - that actually makes a lot of things clear for me. I was able to do a bit of refactoring for our rails engine and make it work in the rails application. We had some concerns in the model's directory starting with `::Concerns` which was bad. I don't remember what error I was getting last time I tried but maybe I was trying too much refactoring in that try.

Now with the information you provided, refactoring became easier. Thanks! Going to close it now but will re-open it if I run into anything.
